lasagne/layers/transform.py

```diff
@@ -0,0 +1,200 @@
+        if len(input_shp) != 4:
+            raise ValueError('The input network must have 4 dimensions: '
+                             '[num_batch, channels, height width]')
+
+    def get_output_shape_for(self, input_shapes):
+        shp = input_shapes[0]
+        dsf = self.downsample_factor
+        return tuple(list(shp[:2]) + [int(s//dsf) for s in shp[2:]])
+
+    def get_output_for(self, inputs, deterministic=False, **kwargs):
+        # theta should be shape (num_batch, 2, 3)
+        # see eq. (1) and sec 3.1 in [1]
+        input, theta = inputs
+        return _transform(theta, input, self.downsample_factor)
+
+
+def _transform(theta, input, downsample_factor):
+    num_batch, num_channels, height, width = input.shape
+    theta = T.reshape(theta, (-1, 2, 3))
+
+    height_f = T.cast(height, 'float32')
+    width_f = T.cast(width, 'float32')
+
+    out_height = T.cast(height_f // downsample_factor, 'int64')
+    out_width = T.cast(width_f // downsample_factor, 'int64')
+
+    # grid of (x_t, y_t, 1), eq (1) in ref [1]
+    grid = _meshgrid(out_height, out_width)
+
+    # Transform A x (x_t, y_t, 1)^T -> (x_s, y_s)
+    T_g = T.dot(theta, grid)
+    x_s = T_g[:, 0]
+    y_s = T_g[:, 1]
+    x_s_flat = x_s.flatten()
+    y_s_flat = y_s.flatten()
+
+    # dimshuffle input to  (bs, height, width, channels)
+    input_dim = input.dimshuffle(0, 2, 3, 1)
+    input_transformed = _interpolate(
+        input_dim, x_s_flat, y_s_flat,
+        downsample_factor)
+
+    output = T.reshape(
+        input_transformed, (num_batch, out_height, out_width, num_channels))
+    output = output.dimshuffle(0, 3, 1, 2)  # dimshuffle to conv format
+    return output
+
+
+def _interpolate(im, x, y, downsample_factor):
+    # *_f are floats
+    num_batch, height, width, channels = im.shape
+    height_f = T.cast(height, 'float32')
+    width_f = T.cast(width, 'float32')
+    out_height = T.cast(height_f // downsample_factor, 'int64')
+    out_width = T.cast(width_f // downsample_factor, 'int64')
+    zero = T.zeros([], dtype='int64')
+    max_y = T.cast(im.shape[1] - 1, 'int64')
+    max_x = T.cast(im.shape[2] - 1, 'int64')
+
+    # scale indices from [-1, 1] to [0, width/height]
+    x = (x + 1.0)*(width_f) / 2.0
+    y = (y + 1.0)*(height_f) / 2.0
+
+    x0 = T.cast(T.floor(x), 'int64')
+    x1 = x0 + 1
+    y0 = T.cast(T.floor(y), 'int64')
+    y1 = y0 + 1
+
+    # note sure if this is nessesary...
+    x0 = T.clip(x0, zero, max_x)
+    x1 = T.clip(x1, zero, max_x)
+    y0 = T.clip(y0, zero, max_y)
+    y1 = T.clip(y1, zero, max_y)
+
+    # The input is [num_batch, height, width, channels]. We do the lookup in
+    # the flattened input, i.e [num_batch*height*width, channels]. We need
+    # to offset all indices to match the flat version
+    dim2 = width
+    dim1 = width*height
+    base = _repeat(
+        T.arange(num_batch, dtype='int32')*dim1, out_height*out_width)
+    base_y0 = base + y0*dim2
+    base_y1 = base + y1*dim2
```

Comment:
Again, this is something that might profit from precomputing in numpy if width and height are fixed at compile time.